### ISSUE TITLE
iOS delivery method does not allow cert_path to be a string

### ISSUE THREAD

**Issue by erbunao** [excid3/noticed#182]
I encountered an issue when a custom `cert_path` method  returns a `StringIO.new("certificate_content_here")` object instead of a file.

If I have this code below and an empty `config/certs/ios/apns.p8` file, the push notification gets delivered without an issue.

```
class SomeNotification
  deliver_by :ios, {
    format:            :ios_format,
    cert_path:         ios_cert_path,
  }
  
  def ios_cert_path
    StringIO.new(Rails.application.credentials.dig(:ios, :apn_certificate).to_s)
  end
end

```

However if the p8 file gets deleted, this exception shows up

```
ArgumentError: Could not find APN cert at '.../config/certs/ios/apns.p8'
from .../noticed-1.5.2/lib/noticed/delivery_methods/ios.rb:12:in `deliver'
```

I believe the code should be able to handle a `cert_path` with a string content and not restrict it from always getting the certificate from a file.

**Comment by excid3** [comment]
StringIO isn't currently supported, but I'd love to! Just hadn't thought about that yet. 😜

Your option also needs to be a symbol so that it calls the method with the matching name (just like an ActiveRecord callback). So you should have:

```ruby
  deliver_by :ios, {
    format:            :ios_format,
    cert_path:         :ios_cert_path,
  }

  def ios_cert_path
  end
```

**Comment by erbunao** [comment]
Ah thanks for clarifying that @excid3!

I wonder how would you manage getting the right cert_path depending on the environment (production vs staging vs development)

**Comment by excid3** [comment]
If you had 
```
config/certs/ios/production.p8
```

**Comment by erbunao** [comment]
Yeah totally! I'm worried to add in the p8 files in the repo as I believe they should be kept in a private place. Therefore I was using Rails credentials to get the content.

Has this file management ever been a concern before?

**Comment by joemasilotti** [comment]
FWIW, I usually commit my p8 files to the repo and put the team ID and key in Rails credentials. That way even if someone got the p8 they couldn't use it unless they also got access to your credentials.

**Comment by erbunao** [comment]
@joemasilotti that makes complete sense! 

Just looping back on the main point, I feel the `cert_path` interface would benefit from having a parity on `Apnotic::Connection`'s interface as well, like so

```
Apnotic::Connection.new(cert_path: StringIO.new("pem cert as string"))
```

Would that be something beneficial from you perspective as well @excid3 @joemasilotti ?

**Comment by excid3** [comment]
All the Apnotic examples in their readme show it with a filename for `cert_path`. Does it actually accept a File object?

**Comment by erbunao** [comment]
Yup that would be ideal @excid3! I can submit a PR for this if you want me to 😄

**Comment by excid3** [comment]
Are there any code changes needed?  Since we just pass the value from that method directly to Apnotic it should work already. I guess just updating the docs with the example?

**Comment by erbunao** [comment]
I was hoping it would be that as well but there is a guard clause in `Noticed::DeliveryMethods::Ios` that checks whether the `cert_path` file exists

```
raise ArgumentError, "Could not find APN cert at '#{cert_path}'" unless File.exist?(cert_path)
```

**Comment by excid3** [comment]
Ah yeah! This easy enough?

```ruby
raise ArgumentError, "Could not find APN cert at '#{cert_path}'" unless cert_path.is_a?(String) && File.exist?(cert_path)
```

Then we can include an example in the docs for `File` or `StringIO` objects. I don't think we really need a validation for those, do you?

**Comment by erbunao** [comment]
Yeah that should be enough I believe. I think there is quite a value to validate the options. We just need to have validation parity with Apnotic's.

**Comment by erbunao** [comment]
@excid3 submitted a PR for it 😄 Let me know what you think!

**Comment by excid3** [comment]
Released in 1.5.5
